File: Python/block_a_byte_unit/byte_encoder.py
"""VRAXION byte encoder (Block A) — Python deploy SDK.

Reads the frozen binary+C19+H=16 champion LUT and provides a minimal API:

  from Python.byte_encoder import ByteEncoder
  enc = ByteEncoder.load_default()
  vec = enc.encode(0x41)          # 16-dim float vector
  bytes_back = enc.decode(vec)    # recover byte

The encoder uses the baked 256-entry int8 LUT. The decoder uses the saved
binary weight matrices (W1, W2) from the winner JSON.

100% lossless on all 256 bytes (verified at champion-freeze time and
reproduced here in tests/test_byte_encoder.py).

Zero ML framework dependency — pure numpy.
"""
from __future__ import annotations
import json
from pathlib import Path
from typing import Union

import numpy as np

DEFAULT_LUT_PATH = (
    Path(__file__).resolve().parent.parent.parent
    / "output" / "byte_unit_champion_binary_c19_h16" / "byte_embedder_lut_int8.json"
)
DEFAULT_WEIGHTS_PATH = (
    Path(__file__).resolve().parent.parent.parent
    / "output" / "byte_unit_champion_binary_c19_h16" / "byte_unit_winner_binary.json"
)


class ByteEncoder:
    """Deploy-ready byte encoder for Block A (binary + C19 + H=16, 100% lossless).

    Two code paths:
      - encode(byte) -> 16-dim vector : O(1) LUT lookup
      - decode(vec)  -> byte          : 1 matmul through tied-mirror weights

    Both are pure numpy. No framework, no autograd, no training here.
    """

    LUT_DIM = 16
    INPUT_BITS = 8
    HIDDEN = 16

    def __init__(self, lut_int8: np.ndarray, lut_scale: float,
                 W1: np.ndarray, W2: np.ndarray):
        assert lut_int8.shape == (256, self.LUT_DIM), f"LUT shape mismatch: {lut_int8.shape}"
        assert lut_int8.dtype == np.int8
        assert W1.shape == (self.INPUT_BITS, self.HIDDEN)
        assert W2.shape == (self.HIDDEN, self.LUT_DIM)
        self._lut_int8 = lut_int8
        self._lut_scale = float(lut_scale)
        self._lut_f32 = lut_int8.astype(np.float32) * self._lut_scale
        self._W1 = W1.astype(np.float32)
        self._W2 = W2.astype(np.float32)

    # ------------------------------------------------------------------
    # factories

    @classmethod
    def load_default(cls) -> "ByteEncoder":
        return cls.from_paths(DEFAULT_LUT_PATH, DEFAULT_WEIGHTS_PATH)

    @classmethod
    def from_paths(cls, lut_path: Path, weights_path: Path) -> "ByteEncoder":
        lut_blob = json.loads(Path(lut_path).read_text(encoding="utf-8"))
        assert lut_blob.get("format") == "int8_lut", f"unexpected LUT format: {lut_blob.get('format')}"
        lut_int8 = np.array(lut_blob["lut"], dtype=np.int8)
        lut_scale = float(lut_blob["scale"])

        w_blob = json.loads(Path(weights_path).read_text(encoding="utf-8"))
        assert w_blob.get("precision") == "binary_scaled", f"unexpected weights precision: {w_blob.get('precision')}"

        # Reconstruct float weights from binary indices + levels (levels already contain alpha scaling).
        W1_idx = np.array(w_blob["W1_binary_idx"], dtype=np.int64)
        W1_levels = np.array(w_blob["W1_levels"], dtype=np.float32)
        W1 = W1_levels[W1_idx]

        W2_idx = np.array(w_blob["W2_binary_idx"], dtype=np.int64)
        W2_levels = np.array(w_blob["W2_levels"], dtype=np.float32)
        W2 = W2_levels[W2_idx]

        return cls(lut_int8=lut_int8, lut_scale=lut_scale, W1=W1, W2=W2)

    # ------------------------------------------------------------------
    # encoding

    def encode(self, byte: Union[int, bytes]) -> np.ndarray:
        """Byte (int 0..255 or 1-byte bytes) -> (16,) float32 latent vector.

        O(1) LUT lookup.
        """
        if isinstance(byte, (bytes, bytearray)):
            assert len(byte) == 1, f"expected 1-byte input, got {len(byte)}"
            b = byte[0]
        else:
            b = int(byte)
        assert 0 <= b < 256, f"byte out of range: {b}"
        return self._lut_f32[b].copy()

    def encode_bytes(self, data: bytes) -> np.ndarray:
        """Vectorized: bytes -> (N, 16) float32 latent matrix."""
        arr = np.frombuffer(data, dtype=np.uint8)
        return self._lut_f32[arr].copy()

    # ------------------------------------------------------------------
    # decoding (tied-mirror)

    def decode(self, latent: np.ndarray) -> int:
        """(16,) latent vector -> byte (int 0..255).

        Runs the tied-mirror decoder: latent @ W2.T @ W1.T -> 8-dim, sign -> bits.
        """
        latent = np.asarray(latent, dtype=np.float32).reshape(-1)
        assert latent.shape == (self.LUT_DIM,), f"latent shape mismatch: {latent.shape}"
        # Decoder: project 16 -> 8 via tied mirror (W2.T then W1.T)
        h = latent @ self._W2.T       # (16,) -> (16,)
        x_hat = h @ self._W1.T        # (16,) -> (8,)
        # Signs of the 8 outputs are the byte bits (bipolar representation)
        bits = (x_hat > 0).astype(np.uint8)
        # Pack bits lowest-first (bit 0 is LSB)
        byte = 0
        for i in range(8):
            byte |= int(bits[i]) << i
        return byte

    def decode_bytes(self, latents: np.ndarray) -> bytes:
        """Vectorized: (N, 16) -> N-byte sequence."""
        latents = np.asarray(latents, dtype=np.float32)
        assert latents.ndim == 2 and latents.shape[1] == self.LUT_DIM
        h = latents @ self._W2.T      # (N, 16)
        x_hat = h @ self._W1.T        # (N, 8)
        bits = (x_hat > 0).astype(np.uint8)
        bytes_out = np.zeros(latents.shape[0], dtype=np.uint8)
        for i in range(8):
            bytes_out |= bits[:, i] << i
        return bytes_out.tobytes()

    # ------------------------------------------------------------------
    # diagnostics

    def verify_lossless(self) -> tuple[int, int]:
        """Encode all 256 bytes, decode them back, return (matches, total)."""
        latents = self._lut_f32  # (256, 16)
        out = self.decode_bytes(latents)
        orig = bytes(range(256))
        matches = sum(1 for a, b in zip(orig, out) if a == b)
        return matches, 256

    def __repr__(self) -> str:
        return (f"ByteEncoder(champion=binary+C19+H=16, "
                f"input=8bits, latent={self.LUT_DIM}dim, "
                f"lut_scale={self._lut_scale:.6e})")


if __name__ == "__main__":
    enc = ByteEncoder.load_default()
    print(enc)
    matches, total = enc.verify_lossless()
    print(f"Self-verify round-trip: {matches}/{total} bytes lossless")

    demo_byte = ord("A")  # 65
    vec = enc.encode(demo_byte)
    print(f"\nencode({demo_byte}, 'A') -> shape={vec.shape}, first 4 = {vec[:4]}")
    back = enc.decode(vec)
    print(f"decode(vec) -> {back} ('{chr(back)}')")

File: Rust/src/block_a_byte_unit/byte_encoder.rs
//! VRAXION byte encoder (Block A) — Rust deploy SDK.
//!
//! Reads the frozen binary+C19+H=16 champion LUT and provides:
//!
//! ```ignore
//! use vraxion::ByteEncoder;
//! let enc = ByteEncoder::load_default()?;
//! let vec = enc.encode(0x41);            // 16-dim f32
//! let byte_back = enc.decode(&vec);      // recover byte
//! ```
//!
//! The encoder uses the baked 256-entry int8 LUT. The decoder uses the saved
//! binary weight matrices (W1, W2) from the winner JSON.
//!
//! 100% lossless on all 256 bytes. Zero ML framework dependency — pure Rust.

use std::fs;
use std::path::{Path, PathBuf};

use serde::Deserialize;

pub const LUT_DIM: usize = 16;
pub const INPUT_BITS: usize = 8;
pub const HIDDEN: usize = 16;

#[derive(Debug)]
pub enum LoadError {
    Io(std::io::Error),
    Parse(serde_json::Error),
    BadShape(&'static str),
    BadFormat(String),
}

impl std::fmt::Display for LoadError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Io(e) => write!(f, "io: {e}"),
            Self::Parse(e) => write!(f, "parse: {e}"),
            Self::BadShape(m) => write!(f, "bad shape: {m}"),
            Self::BadFormat(m) => write!(f, "bad format: {m}"),
        }
    }
}

impl std::error::Error for LoadError {}

impl From<std::io::Error> for LoadError {
    fn from(e: std::io::Error) -> Self {
        Self::Io(e)
    }
}

impl From<serde_json::Error> for LoadError {
    fn from(e: serde_json::Error) -> Self {
        Self::Parse(e)
    }
}

#[derive(Deserialize)]
struct LutBlob {
    format: String,
    scale: f32,
    lut: Vec<Vec<i8>>,
}

#[derive(Deserialize)]
struct WeightsBlob {
    precision: String,
    #[serde(rename = "W1_binary_idx")]
    w1_idx: Vec<Vec<i64>>,
    #[serde(rename = "W1_levels")]
    w1_levels: Vec<f32>,
    #[serde(rename = "W2_binary_idx")]
    w2_idx: Vec<Vec<i64>>,
    #[serde(rename = "W2_levels")]
    w2_levels: Vec<f32>,
}

/// Deploy-ready byte encoder for Block A (binary + C19 + H=16, 100% lossless).
pub struct ByteEncoder {
    /// Decoded float LUT (256 × LUT_DIM). Row `b` is the latent for byte `b`.
    lut_f32: [[f32; LUT_DIM]; 256],
    /// W1 weights (INPUT_BITS × HIDDEN), already multiplied by alpha1.
    w1: [[f32; HIDDEN]; INPUT_BITS],
    /// W2 weights (HIDDEN × LUT_DIM), already multiplied by alpha2.
    w2: [[f32; LUT_DIM]; HIDDEN],
}

impl ByteEncoder {
    /// Load from the repo's default champion paths.
    ///
    /// Resolves relative to the crate directory (`CARGO_MANIFEST_DIR/..`) so
    /// `cargo test` works from any working directory. For runtime deploy that
    /// ships the champion artifacts next to the binary, use [`Self::from_paths`].
    pub fn load_default() -> Result<Self, LoadError> {
        let repo_root = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .expect("Rust/ parent")
            .to_path_buf();
        let base = repo_root.join("output").join("byte_unit_champion_binary_c19_h16");
        Self::from_paths(
            base.join("byte_embedder_lut_int8.json"),
            base.join("byte_unit_winner_binary.json"),
        )
    }

    /// Load from explicit paths.
    pub fn from_paths<P: AsRef<Path>, Q: AsRef<Path>>(
        lut_path: P,
        weights_path: Q,
    ) -> Result<Self, LoadError> {
        let lut_text = fs::read_to_string(lut_path.as_ref())?;
        let lut: LutBlob = serde_json::from_str(&lut_text)?;
        if lut.format != "int8_lut" {
            return Err(LoadError::BadFormat(format!(
                "expected int8_lut, got {}",
                lut.format
            )));
        }
        if lut.lut.len() != 256 {
            return Err(LoadError::BadShape("LUT rows != 256"));
        }
        if lut.lut[0].len() != LUT_DIM {
            return Err(LoadError::BadShape("LUT cols != 16"));
        }

        let mut lut_f32 = [[0f32; LUT_DIM]; 256];
        for (b, row) in lut.lut.iter().enumerate() {
            for (j, &v) in row.iter().enumerate() {
                lut_f32[b][j] = (v as f32) * lut.scale;
            }
        }

        let w_text = fs::read_to_string(weights_path.as_ref())?;
        let w: WeightsBlob = serde_json::from_str(&w_text)?;
        if w.precision != "binary_scaled" {
            return Err(LoadError::BadFormat(format!(
                "expected binary_scaled weights, got {}",
                w.precision
            )));
        }
        if w.w1_idx.len() != INPUT_BITS || w.w1_idx[0].len() != HIDDEN {
            return Err(LoadError::BadShape("W1 shape != (8, 16)"));
        }
        if w.w2_idx.len() != HIDDEN || w.w2_idx[0].len() != LUT_DIM {
            return Err(LoadError::BadShape("W2 shape != (16, 16)"));
        }

        let mut w1 = [[0f32; HIDDEN]; INPUT_BITS];
        for i in 0..INPUT_BITS {
            for j in 0..HIDDEN {
                let idx = w.w1_idx[i][j] as usize;
                w1[i][j] = w.w1_levels[idx];
            }
        }
        let mut w2 = [[0f32; LUT_DIM]; HIDDEN];
        for i in 0..HIDDEN {
            for j in 0..LUT_DIM {
                let idx = w.w2_idx[i][j] as usize;
                w2[i][j] = w.w2_levels[idx];
            }
        }

        Ok(Self { lut_f32, w1, w2 })
    }

    /// Byte (0..=255) -> 16-dim f32 latent. O(1) LUT lookup.
    pub fn encode(&self, byte: u8) -> [f32; LUT_DIM] {
        self.lut_f32[byte as usize]
    }

    /// Vectorized encode: bytes -> matrix (one row per byte).
    pub fn encode_bytes(&self, data: &[u8]) -> Vec<[f32; LUT_DIM]> {
        data.iter().map(|&b| self.lut_f32[b as usize]).collect()
    }

    /// (16,) latent -> byte via tied-mirror decode: latent @ W2^T @ W1^T -> signs -> bits.
    pub fn decode(&self, latent: &[f32; LUT_DIM]) -> u8 {
        // Step 1: h = latent @ W2^T, shape (HIDDEN,)
        let mut h = [0f32; HIDDEN];
        for k in 0..HIDDEN {
            let mut acc = 0f32;
            for j in 0..LUT_DIM {
                acc += latent[j] * self.w2[k][j];
            }
            h[k] = acc;
        }

        // Step 2: x_hat = h @ W1^T, shape (INPUT_BITS,)
        let mut x_hat = [0f32; INPUT_BITS];
        for i in 0..INPUT_BITS {
            let mut acc = 0f32;
            for k in 0..HIDDEN {
                acc += h[k] * self.w1[i][k];
            }
            x_hat[i] = acc;
        }

        // Step 3: pack signs into byte (bit i = bit position i)
        let mut byte: u8 = 0;
        for i in 0..INPUT_BITS {
            if x_hat[i] > 0.0 {
                byte |= 1 << i;
            }
        }
        byte
    }

    /// Vectorized decode: latents matrix -> byte sequence.
    pub fn decode_bytes(&self, latents: &[[f32; LUT_DIM]]) -> Vec<u8> {
        latents.iter().map(|l| self.decode(l)).collect()
    }

    /// Encode all 256 bytes, decode them back, return (matches, 256).
    pub fn verify_lossless(&self) -> (usize, usize) {
        let mut matches = 0;
        for b in 0u16..256 {
            let vec = self.lut_f32[b as usize];
            let out = self.decode(&vec);
            if out == b as u8 {
                matches += 1;
            }
        }
        (matches, 256)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn load_for_test() -> ByteEncoder {
        ByteEncoder::load_default().expect("load champion")
    }

    #[test]
    fn default_load_self_verifies_lossless() {
        let enc = load_for_test();
        let (m, t) = enc.verify_lossless();
        assert_eq!((m, t), (256, 256));
    }

    #[test]
    fn encode_shape_is_16_floats() {
        let enc = load_for_test();
        let v = enc.encode(0x41);
        assert_eq!(v.len(), 16);
    }

    #[test]
    fn round_trip_all_256_bytes() {
        let enc = load_for_test();
        for b in 0u16..256 {
            let v = enc.encode(b as u8);
            let back = enc.decode(&v);
            assert_eq!(back, b as u8, "byte {b} round-tripped to {back}");
        }
    }

    #[test]
    fn vectorized_round_trip_ascii() {
        let enc = load_for_test();
        let text = b"The quick brown fox jumps over the lazy dog. 0123456789!@#$%^&*()";
        let latents = enc.encode_bytes(text);
        let back = enc.decode_bytes(&latents);
        assert_eq!(back, text);
    }

    #[test]
    fn vectorized_round_trip_utf8() {
        let enc = load_for_test();
        let text = "Péter szépen éneklő bárány 中文 العربية 🐈".as_bytes();
        let latents = enc.encode_bytes(text);
        let back = enc.decode_bytes(&latents);
        assert_eq!(back, text);
    }
}
